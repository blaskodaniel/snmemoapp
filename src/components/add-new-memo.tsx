import React, { useState } from 'react'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import { Grid } from '@material-ui/core'
import Typography from '@material-ui/core/Typography'
import Fab from '@material-ui/core/Fab'
import Paper from '@material-ui/core/Paper'
import SaveIcon from '@material-ui/icons/Save'
import ClearIcon from '@material-ui/icons/Clear'
import TextField from '@material-ui/core/TextField'
import { AddNewprops } from '../interfaces'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    buttonDefault: {
      margin: theme.spacing(1),
    },
    fab: {
      margin: theme.spacing(1),
    },
    paper: {
      padding: '30px',
      marginBottom: '10px',
    },
  }),
)

export const AddNew: React.FunctionComponent<AddNewprops> = props => {
  const classes = useStyles()
  const [displayname, setDisplayname] = useState('')
  const [description, setDescritption] = useState('')

  const create = () => {
    if (displayname && description) {
      props.onCreate({ DisplayName: displayname, Description: description })
      setDisplayname('')
      setDescritption('')
    }
  }

  return (
    <Paper className={classes.paper} style={{ display: props.show ? 'block' : 'none' }}>
      <Grid container spacing={2}>
        <Grid item xs={12}>
          <Typography variant="h5" component="h3">
            Add new memo to list
          </Typography>
        </Grid>
        <Grid item xs={12} md={6}>
          <TextField
            style={{ width: '100%' }}
            required
            multiline={true}
            label="Memo title"
            value={displayname}
            onChange={ev => setDisplayname(ev.target.value)}
            rows={1}
            rowsMax={10}
            id={'DisplayName'}
          />
        </Grid>
        <Grid item xs={12}>
          <TextField
            style={{ width: '100%' }}
            required
            multiline={true}
            label="Memo content"
            value={description}
            onChange={ev => setDescritption(ev.target.value)}
            rows={1}
            rowsMax={10}
            id={'Description'}
          />
        </Grid>
        <Grid item xs={12} style={{ textAlign: 'right' }}>
          <Fab
            aria-label="Create"
            size={'small'}
            style={{ display: displayname && description ? 'inline-flex' : 'none' }}
            onClick={create}
            className={classes.fab}>
            <SaveIcon />
          </Fab>
          <Fab aria-label="Cancel" size={'small'} onClick={props.onClose} className={classes.fab}>
            <ClearIcon />
          </Fab>
        </Grid>
      </Grid>
    </Paper>
  )
}
