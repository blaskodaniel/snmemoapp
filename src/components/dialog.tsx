import React from 'react'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import DialogTitle from '@material-ui/core/DialogTitle'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import DialogContent from '@material-ui/core/DialogContent'
import DialogContentText from '@material-ui/core/DialogContentText'
import Button from '@material-ui/core/Button'

export interface DialogProps {
  open: boolean
  title: string
  onClose: (value: boolean) => void
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    buttonDefault: {
      margin: theme.spacing(1),
    },
  }),
)

export const DialogComponent: React.FunctionComponent<DialogProps> = props => {
  const classes = useStyles()

  return (
    <Dialog open={props.open} maxWidth="sm" aria-labelledby="simple-dialog-title">
      <DialogTitle id="simple-dialog-title">Are you sure you want to delete it?</DialogTitle>
      <DialogContent>
        <DialogContentText id="alert-dialog-description">{props.title}</DialogContentText>
      </DialogContent>
      <DialogActions>
        <Button onClick={() => props.onClose(true)} color="secondary" className={classes.buttonDefault}>
          Yes
        </Button>
        <Button onClick={() => props.onClose(false)} color="default" className={classes.buttonDefault}>
          No
        </Button>
      </DialogActions>
    </Dialog>
  )
}
