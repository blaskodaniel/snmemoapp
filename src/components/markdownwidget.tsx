import React from 'react'
import { Typography } from '@material-ui/core'
import ReactMarkdown from 'react-markdown'
import { MarkdownWidgetI } from '../interfaces'

export const MarkdownWidget: React.FunctionComponent<MarkdownWidgetI> = props => {
  return (
    <div>
      <Typography
        variant="h5"
        title={props.title}
        gutterBottom={true}
        style={{ whiteSpace: 'nowrap', textOverflow: 'ellipsis', overflow: 'hidden' }}>
        {props.title}
      </Typography>
      <div style={{ overflow: 'auto' }}>
        <ReactMarkdown escapeHtml={false} source={props.settings.content} />
      </div>
    </div>
  )
}
