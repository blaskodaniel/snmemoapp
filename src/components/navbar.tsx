import React from 'react'
import { AppBar, Button, Toolbar, Tooltip, Typography } from '@material-ui/core'
import { createStyles, makeStyles } from '@material-ui/core/styles'
import { useRepository } from '@sensenet/hooks-react'

const useStyles = makeStyles(() =>
  createStyles({
    title: {
      flexGrow: 1,
    },
  }),
)

export const NavBarComponent: React.FunctionComponent = () => {
  const repo = useRepository()
  const classes = useStyles()
  return (
    <AppBar position="fixed">
      <Toolbar>
        <Typography variant="h6" className={classes.title}>
          Memo application
        </Typography>
        <Tooltip title="Return to the Login screen and select another repository">
          <Button color="inherit" onClick={() => repo.authentication.logout()}>
            Log out
          </Button>
        </Tooltip>
      </Toolbar>
    </AppBar>
  )
}
