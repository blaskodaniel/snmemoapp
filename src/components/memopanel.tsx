import React, { useEffect, useState } from 'react'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import moment from 'moment'
import Fab from '@material-ui/core/Fab'
import DeleteIcon from '@material-ui/icons/Delete'
import EditIcon from '@material-ui/icons/Edit'
import SaveIcon from '@material-ui/icons/Save'
import RedoIcon from '@material-ui/icons/Redo'
import ExpansionPanel from '@material-ui/core/ExpansionPanel'
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails'
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary'
import TextField from '@material-ui/core/TextField'
import Typography from '@material-ui/core/Typography'
import ExpandMoreIcon from '@material-ui/icons/ExpandMore'
import ReactMarkdown from 'react-markdown'
import { Grid } from '@material-ui/core'
import { ConstantContent, ODataCollectionResponse } from '@sensenet/client-core'
import { GenericContent, User } from '@sensenet/default-content-types'
import { useRepository } from '../hooks/use-repository'
import { DialogComponent } from '../components/dialog'

const useStyles: any = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      width: '100%',
    },
    hidden: {
      display: 'none',
    },
    heading: {
      fontSize: theme.typography.pxToRem(15),
      flexBasis: '33.33%',
      flexShrink: 0,
    },
    secondaryHeading: {
      fontSize: theme.typography.pxToRem(15),
      color: theme.palette.text.secondary,
    },
    fab: {
      margin: theme.spacing(1),
    },
    extendedIcon: {
      marginRight: theme.spacing(1),
    },
  }),
)

export const MemoPanel: React.FunctionComponent = () => {
  const repo: any = useRepository()
  const classes: any = useStyles()
  const [expanded, setExpanded] = useState<string | false>(false)
  const [editmode, setEditmode] = useState<string | false>('')
  const [editText, setEditText] = useState<string>('')
  const [openmodal, setOpenmodal] = useState<boolean>(false)
  const [modaltitle, setModaltitle] = useState<string>('')
  const [currentmemo, setCurrentmemo] = useState<GenericContent>(null as any)
  const [data, setData] = useState<GenericContent[]>([])

  useEffect(() => {
    async function loadMemos(): Promise<void> {
      const result: ODataCollectionResponse<GenericContent> = await repo.loadCollection({
        path: `${ConstantContent.PORTAL_ROOT.Path}/Content/IT/Memos`,
        oDataOptions: {
          select: ['DisplayName', 'Description', 'CreationDate', 'CreatedBy', 'ModificationDate'] as any,
          orderby: [['ModificationDate', 'desc']],
          expand: ['CreatedBy'] as string[],
        },
      })

      setData(result.d.results)
    }
    loadMemos()
  }, [repo])

  // Remove memo - open modal
  const deleteOpenModal = (memo: GenericContent) => {
    setCurrentmemo(memo) // set current content for removing
    const title = memo.DisplayName ? memo.DisplayName : ''
    setModaltitle(title)
    setOpenmodal(true)
  }

  // Remove memo - remove content
  const deleteMemoContent = async (receivebtn: boolean) => {
    if (receivebtn) {
      // User clicked "Yes"
      const newdata = data.filter(x => x.Id != currentmemo.Id)
      await repo.delete({
        idOrPath: currentmemo.Path,
        permanent: true,
      })
      setData(newdata) // refresh memo list
      setOpenmodal(false) // close modal window
    } else {
      // User clicked "No"
      setOpenmodal(false)
    }
  }

  // Expansion panel handler
  const handleChangeExpand: any = (panel: string) => (_event: React.ChangeEvent<{}>, isExpanded: boolean) => {
    setExpanded(isExpanded ? panel : false)
    console.log(expanded)
  }

  // Edit/Read mode handler
  const handleEditmode: any = (memo: GenericContent) => () => {
    setEditmode(memo.Id.toString())
    setEditText(memo.Description ? memo.Description : '')
    console.log(editmode)
  }

  // Save handler
  const handleSave: any = async (memo: GenericContent) => {
    const newDescrition = {
      Description: editText,
    }
    await repo.patch({
      idOrPath: memo.Id,
      content: newDescrition,
    })
    data.map(x => {
      if (x.Id === memo.Id) {
        x.Description = editText
      }
    })
    setData(data)
    setEditmode(false)
    console.log('Save successfuly')
  }

  // Text change handler in edit mode
  const handleChange: any = () => (event: React.ChangeEvent<HTMLInputElement>) => {
    setEditText(event.target.value)
  }

  return (
    <div className={classes.root}>
      {data.map(memo => (
        <ExpansionPanel
          key={memo.Id}
          expanded={expanded === memo.Id.toString()}
          onChange={handleChangeExpand(memo.Id.toString())}>
          <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />} aria-controls="panel1bh-content" id="panel1bh-header">
            <Typography className={classes.heading}>{memo.DisplayName}</Typography>
            <Typography className={classes.secondaryHeading}>
              {memo.CreatedBy ? (memo.CreatedBy as User).FullName : ''} -{' '}
              {moment(new Date(memo.CreationDate ? memo.CreationDate : '')).format('YYYY-MM-DD HH:mm:ss')}
            </Typography>
          </ExpansionPanelSummary>
          <ExpansionPanelDetails>
            <Grid container>
              <Grid item xs={12}>
                <ReactMarkdown
                  source={memo.Description}
                  className={editmode === memo.Id.toString() ? classes.hidden : ''}
                />
                <TextField
                  style={{ width: '100%', display: editmode === memo.Id.toString() ? 'block' : 'none' }}
                  placeholder="Write a memo..."
                  multiline={true}
                  value={editText}
                  onChange={handleChange(memo.Id.toString())}
                  rows={1}
                  rowsMax={10}
                  id={memo.Id.toString()}
                />
              </Grid>
              <Grid item xs={12} style={{ textAlign: 'right' }}>
                <Fab
                  color="secondary"
                  aria-label="Edit"
                  size={'small'}
                  className={editmode === memo.Id.toString() ? classes.hidden : classes.fab}>
                  <EditIcon onClick={handleEditmode(memo)} />
                </Fab>
                <Fab
                  aria-label="Delete"
                  size={'small'}
                  onClick={() => deleteOpenModal(memo)}
                  className={editmode === memo.Id.toString() ? classes.hidden : classes.fab}>
                  <DeleteIcon />
                </Fab>
                <Fab
                  aria-label="Save"
                  size={'small'}
                  onClick={() => handleSave(memo)}
                  className={editmode === memo.Id.toString() ? classes.fab : classes.hidden}>
                  <SaveIcon />
                </Fab>
                <Fab
                  aria-label="Cancel"
                  size={'small'}
                  onClick={() => {
                    setEditmode(false)
                  }}
                  className={editmode === memo.Id.toString() ? classes.fab : classes.hidden}>
                  <RedoIcon />
                </Fab>
              </Grid>
            </Grid>
          </ExpansionPanelDetails>
        </ExpansionPanel>
      ))}
      <DialogComponent
        open={openmodal}
        title={modaltitle}
        onClose={receivebtn => {
          deleteMemoContent(receivebtn)
        }}
      />
    </div>
  )
}
