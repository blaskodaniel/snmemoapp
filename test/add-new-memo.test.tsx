import React from 'react'
import { mount, shallow } from 'enzyme'
import { TextField } from '@material-ui/core'
import { act } from 'react-dom/test-utils'
import { AddNew } from '../src/components/add-new-memo'

describe('The new memo panel instance', () => {
  const addnewprops = {
    show: true,
    onCreate: jest.fn(),
    onClose: jest.fn(),
  }
  it('should renders correctly', () => {
    const wrapper = shallow(<AddNew {...addnewprops} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should create a memo', () => {
    const wrapper = mount(<AddNew {...addnewprops} />)
    const textfieldTitle = wrapper.find(TextField).at(0)
    const textfieldDesc = wrapper.find(TextField).at(1)

    act(() => {
      ;(textfieldTitle.prop('onChange') as any)({ target: { value: 'New memo title' } })
    })

    act(() => {
      ;(textfieldDesc.prop('onChange') as any)({ target: { value: 'New memo description' } })
    })

    expect(
      textfieldTitle
        .find('textarea')
        .at(0)
        .text(),
    ).toEqual('New memo title')
    expect(
      textfieldDesc
        .find('textarea')
        .at(0)
        .text(),
    ).toEqual('New memo description')

    act(() => {
      wrapper
        .update()
        .find('button[aria-label="Create"]')
        .simulate('click')
    })

    expect(
      textfieldTitle
        .find('textarea')
        .at(0)
        .text(),
    ).toEqual('')
    expect(
      textfieldDesc
        .find('textarea')
        .at(0)
        .text(),
    ).toEqual('')
  })
})
