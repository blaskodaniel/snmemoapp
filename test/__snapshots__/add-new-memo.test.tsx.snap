// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The new memo panel instance should renders correctly 1`] = `
<WithStyles(ForwardRef(Paper))
  className="makeStyles-paper-3"
  style={
    Object {
      "display": "block",
    }
  }
>
  <WithStyles(ForwardRef(Grid))
    container={true}
    spacing={2}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={12}
    >
      <WithStyles(ForwardRef(Typography))
        component="h3"
        variant="h5"
      >
        Add new memo to list
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      md={6}
      xs={12}
    >
      <WithStyles(ForwardRef(TextField))
        id="DisplayName"
        label="Memo title"
        multiline={true}
        onChange={[Function]}
        required={true}
        rows={1}
        rowsMax={10}
        style={
          Object {
            "width": "100%",
          }
        }
        value=""
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={12}
    >
      <WithStyles(ForwardRef(TextField))
        id="Description"
        label="Memo content"
        multiline={true}
        onChange={[Function]}
        required={true}
        rows={1}
        rowsMax={10}
        style={
          Object {
            "width": "100%",
          }
        }
        value=""
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      style={
        Object {
          "textAlign": "right",
        }
      }
      xs={12}
    >
      <WithStyles(ForwardRef(Fab))
        aria-label="Create"
        className="makeStyles-fab-2"
        onClick={[Function]}
        size="small"
        style={
          Object {
            "display": "none",
          }
        }
      >
        <SaveIcon />
      </WithStyles(ForwardRef(Fab))>
      <WithStyles(ForwardRef(Fab))
        aria-label="Cancel"
        className="makeStyles-fab-2"
        onClick={[MockFunction]}
        size="small"
      >
        <ClearIcon />
      </WithStyles(ForwardRef(Fab))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Paper))>
`;
